### DISCUSSION TITLE
can't modify atom in a loop, can't access newest state without re-render

### DISCUSSION THREAD

**Discussion by likon416** [facebookexperimental/Recoil#1686] [Q&A]
If we have an atom and we want to for example do something like
const [myAtom, setMyAtom] = atom(defaultValue);

for(let i=0; i < 10; i++) {
setMyAtom(myAtom + 1);
}

It will not increase the atom's number by 10 but by 1. Because on write, atom refreshes, the next loop iteration, doesn't see the new atom state (because component didn't re-render yet with new atom value). 

I know you can write the code differently and just do setAtom(myAtom + 10) but this prevent us from writing code like
removeItemFromList();
addItemFromList();

Because removing and adding would use setMyAtom inside and they are executed one after another, just one operation will be written(because between executing those methods, component didn't re-render). This really limits how we can write the code, it's not flexible. You can do such operations in Redux but not in atoms.

1. Atom should always see the newest state, regardless of how many times we will do write on it, it can always notify to refresh with some delay if needed. 
2. Atom should always have 1 singleton instance so whenever you will do write or read it always has the newest state.
3. Atom newest state should always be accessible without need to re-render. Re-render should be just for notifying that state changed.

Is it currently possible to achieve the above?

Thanks :)

**Comment by jeffscalyr** [comment]
I think the most straightforward way to achieve what you want is to pass an updater function to your setter.
From the docs for `useSetRecoilState`:

_The setter may either be passed a new value or an updater function which receives the previous value as an argument._

Check out how `setNamesState` is passed a function in the example from the same docs for `useSetRecoilState`:

```
import {atom, useSetRecoilState} from 'recoil';

const namesState = atom({
  key: 'namesState',
  default: ['Ella', 'Chris', 'Paul'],
});

function FormContent({setNamesState}) {
  const [name, setName] = useState('');
  
  return (
    <>
      <input type="text" value={name} onChange={(e) => setName(e.target.value)} />
      <button onClick={() => setNamesState(names => [...names, name])}>Add Name</button>
    </>
)}

// This component will be rendered once when mounting
function Form() {
  const setNamesState = useSetRecoilState(namesState);
  
  return <FormContent setNamesState={setNamesState} />;
}
```

**Comment by likon416** [comment]
@jeffscalyr looks interesting, will try this one tomorrow :) thanks!

**Reply by likon416** [reply]
@jeffscalyr that worked as well, perfect :) thanks!

**Reply by likon416** [reply]
@jeffscalyr when using such a loop that will increase numbers 10 times, and our atom is using effect to save data to the storage, will it save to the storage 10 times, or it will wait for the end of the loop and save to the storage just once?

**Reply by jeffscalyr** [reply]
I think that Recoil state updates within the same synchronous chunk of JavaScript are batched (as are the underlying React updates).

I'm not sure whether the same holds true for atom effects; I would guess it does.

I am curious, can you write your code in this form?

```
    setMyAtom((value) => {
      for (let i = 0; i < 10; i++) {
        value += 1;
      }
      return value;
    });
```

I think I would use the above format when possible for readability and style anyway.

If you really do want to call one or more setters in sequence and have it treated as one atomic transaction, I think `useRecoilTransaction_UNSTABLE` might be what you want.

**Comment by likon416** [comment]
@jeffscalyr yeah it looks like it's batching those changes and running effect only once. 
This what you wrote is an alternative way to write it. But imagine you want to write a dynamic list modifier. Which dynamically, removes some items, add them, or maybe change their names. Would be good to be able to write a loop that will do all necessary add, remove, changeName. Even if you would put it inside setMyAtom, in my opinion it's very important to be able to reuse code and separate some bits to separate methods like add, remove, changeName etc. so you can use them freely, without need to duplicate the code. Fortunately setMyAtom(data => return val) seems to do the trick :)
